cart/test/api.test.js

```diff
@@ -0,0 +1,213 @@
+const requestCart = require("supertest")(cartURL);
+const logger = require('../src/utils/logger');
+
+describe("Cart REST API", () => {
+  let product, cartid, cart, sample_products, sample_users;
+
+  // Utility function to initialize data
+  async function loadSampleData() {
+    // Set up test data
+    sample_users = [
+      {
+        "uid": "user1",
+        "fname": "John",
+        "lname": "Doe",
+        "address": "Some Street 22",
+        "phone": 1231231231,
+        "email": "john@doe.com"
+      }
+    ]
+
+    sample_products = [
+      {
+        "pid": 1,
+        "pcode": "ABC123",
+        "price": 42.5,
+        "sku": "XYZ",
+        "amount_in_stock": 10,
+        "pname": "Something",
+        "description": "This is something very interesting that you want to buy.",
+        "lang": "en_US"
+      },
+      {
+        "pid": 2,
+        "pcode": "FD2",
+        "price": 99.99,
+        "sku": "QWOP",
+        "amount_in_stock": 100,
+        "pname": "Speedos",
+        "description": "I'm too lazy to write a description",
+        "lang": "en_US"
+      }
+    ]
+
+    // Create the records in the database
+    logger.debug(`Inserting sample records`);
+    await cart.deleteAll(); // Clear out the existing data
+    await cart.repository.knex('products').insert(sample_products);
+    await cart.repository.knex('users').insert(sample_users);
+  }
+
+  // Runs before all tests
+  before(function before() {
+
+    // Load the config and wait for it to load
+    const config = require('../src/configs/config');
+    config.then(() => {
+
+      // Set up all the required constants
+      const Kernel = require('../src/models/kernel');
+      const Names = require('../src/constants/modelNames');
+      const knexManager = Kernel.resolve(Names.knexManager);
+      const knexInstance = knexManager.knex
+      const Pg = require('../src/models/repository');
+      const Cart = require('../src/models').Model;
+
+      // Set up the cart model to be able to call its methods directly
+      let repoOpts = {knex: knexInstance, resource: 'products_in_cart', logger: logger};
+      let repository = new Pg(repoOpts);
+      cart = new Cart({repository: repository});
+
+      // Log the start of the test
+      logger.debug(`Starting test at ${new Date().toLocaleString()}`);
+
+      // Load the sample data into the database
+      loadSampleData();
+
+      // Define a sample reference to a product for later use
+      product = {
+        pid: 1,
+        amount_in_cart: 3
+      }
+      // Choose a sample cart id
+      cartid = 1;
+    })
+    .catch(err => {
+      logger.error(err.message);
+      throw err;
+    })
+  })
+
+  // Before each test, clear out the cart data
+  beforeEach(async function beforeEach() {
+    await cart.deleteAll();
+  })
+
+  // Test API functionality
+  it("lists products", async () => {
+    // Check if product listing is possible
+    const res = await requestCart.get(`/${cartid}`).expect(200)
```

Comment:
naming convention for cartid throughout the file. should be in camelCase

Reply:
In the database, the column is cartid, because postgres is case-insensitive by default and creates tables in lowercase.

Three options:

1. Make the tables case-sensitive in the database
1. Use lowercase cartid (keep it as it is)
1. Specify {cartid: cartId} in the where query, which goes against one of your previous points.

I prefer option two. Which do you (or others) think is best?

Reply:
The fields or column names in database should follow snake_case. So cartid should be cart_id. All the JavaScript fields should be camelCase. And so: {cart_id: cartId}